shell/platform/windows/windows_lifecycle_manager.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "windows_lifecycle_manager.h"
+
+#include <TlHelp32.h>
+#include <WinUser.h>
+#include <Windows.h>
+#include <tchar.h>
+
+#include "flutter/shell/platform/windows/flutter_windows_engine.h"
+
+namespace flutter {
+
+WindowsLifecycleManager::WindowsLifecycleManager(FlutterWindowsEngine* engine)
+    : engine_(engine) {}
+
+WindowsLifecycleManager::~WindowsLifecycleManager() {}
+
+void WindowsLifecycleManager::Quit(UINT exit_code, HWND hwnd) {
+  if (hwnd == nullptr) {
+    ::PostQuitMessage(exit_code);
+  } else {
+    sent_close_messages_[hwnd]++;
```

Comment:
They keyboard logs an error if there are too many pending redispatches messages as that indicates redispatching is broken. Should we do that as well?

Reply:
For the keyboard, it should be impossible to get consecutive keyup msgs without a keydown in between, right? But I don't think there's any such guarantee for windows, e.g. if the close button is clicked the same instant as alt+f4 is pressed, so I am not so sure this should be an error state.

Reply:
The error is to catch a different scenario: we're redispatching messages but not receiving them. An unbalance would result us in dropping `WM_CLOSE` messages unnecessarily.

Reply:
What would be the condition that you are concerned may arise in this case, that the value held by the map exceeds 1? If so, could we not need to redispatch 2 messages at a time?

Reply:
If we redispatch `WM_CLOSE` messages we expect they'll be handled shortly thereafter. It would be unexpected for us to have many redispatched `WM_CLOSE` messages in flight. This may indicate that our redispatch logic is broken. However, I don't feel strongly about asserting for this scenario.